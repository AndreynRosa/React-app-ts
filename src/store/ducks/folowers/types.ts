/**
 * Action types
 */
export enum FollowersTypes {
  LOAD_REQUEST = '@followers/LOAD_REQUEST',
  LOAD_SUCCCES = '@followers/LOAD_SUCCCES',
  LOAD_FAILURE = '@followers/LOAD_FAILURE',
}

/**
 * Data types
 */
export interface Followers {
  login: string;
  id: number;
}

/**
 * State type
 */
export interface FollowersState {
  readonly data: Followers[];
  readonly loading: boolean;
  readonly error: boolean;
}
