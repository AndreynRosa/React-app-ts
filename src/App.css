@import '~antd/dist/antd.css';



.CnterText{
  text-align: 'center';
}
