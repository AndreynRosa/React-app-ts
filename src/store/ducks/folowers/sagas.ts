import { call, put } from 'redux-saga/effects';

import { ActionType } from 'typesafe-actions';
import { folowersLoadSuccess, folowersLoadFailure } from './actions';
import api from '../../../services/api';
import * as actions from './actions';

export function* loadFlowers(
  action: ActionType<typeof actions.folowersLoadRequest>,
) {
  try {
    const url = `users/${action.payload.name}/followers`;
    const response = yield call(api.get, url);
    yield put(folowersLoadSuccess(response.data));
  } catch (err) {
    yield put(folowersLoadFailure());
  }
}
