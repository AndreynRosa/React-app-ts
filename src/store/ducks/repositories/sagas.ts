import { call, put } from 'redux-saga/effects';

import { ActionType } from 'typesafe-actions';
import { repositoriesLoadSuccess, repositoriesLoadFailure } from './actions';
import api from '../../../services/api';
import * as actions from './actions';

export function* loadRepositories(
  action: ActionType<typeof actions.repositoriesLoadRequest>,
) {
  try {
    const url = `users/${action.payload.name}/repos`;
    console.log('url', url);
    const response = yield call(api.get, url);

    yield put(repositoriesLoadSuccess(response.data));
  } catch (err) {
    yield put(repositoriesLoadFailure());
  }
}
