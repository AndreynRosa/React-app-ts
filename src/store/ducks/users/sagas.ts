import { call, put } from 'redux-saga/effects';
import { ActionType } from 'typesafe-actions';

import { userLoadSuccess, userLoadFailure } from './actions';
import api from '../../../services/api';
import * as actions from './actions';
import { Users } from './types';

interface Organization {
  url: string;
}
async function getOrganizationNames(organizationsURL: string) {
  const response = await api.get(organizationsURL);
  return response.data?.name;
}
export function* loadUser(action: ActionType<typeof actions.userLoadRequest>) {
  try {
    const response = yield call(api.get, `users/${action.payload.name}`);
    const organizationsReponse = yield call(
      api.get,
      `users/${action.payload.name}/orgs`,
    );
    const organizations = organizationsReponse.data.map(
      (organization: Organization) => {
        return organization.url;
      },
    );
    const organization01 = yield call(getOrganizationNames, organizations[0]);
    const organization02 = yield call(getOrganizationNames, organizations[1]);

    const { name, following, email, public_gists, avatar_url } = response.data;
    const buildRep: Users[] = [
      {
        name,
        avatar: avatar_url,
        following,
        email,
        organizations: [organization01, organization02],
        public_gists,
      },
    ];
    yield put(userLoadSuccess(buildRep));
  } catch (err) {
    yield put(userLoadFailure());
  }
}
