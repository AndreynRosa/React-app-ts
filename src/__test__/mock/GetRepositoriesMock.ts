import { Repository } from "../../store/ducks/repositories/types";
import GetRepositoryMock from "./GetRepositoryMock";

const GetRepositoriesMock: Repository[] = [GetRepositoryMock];
export default GetRepositoriesMock;
