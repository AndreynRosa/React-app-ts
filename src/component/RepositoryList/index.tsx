import React from 'react';
import { useSelector } from 'react-redux';

import { List, Typography } from 'antd';

import { RepositoriesState } from '../../store/ducks/repositories/types';

interface RootState {
  repositories: RepositoriesState;
}

const RepositoryList: React.FC = () => {
  const repositoryList = useSelector((state: RootState) => state.repositories);
  const { Title, Text } = Typography;

  return (
    <>
      <List
        style={{ height: '244px', background: '#fff' }}
        size="small"
        header={<Title level={4}>Repository</Title>}
        bordered
        dataSource={repositoryList.data}
        renderItem={(item) => (
          <List.Item>
            <Text>
              {item.name}
              {item.id}
            </Text>
          </List.Item>
        )}
      />
    </>
  );
};

export default RepositoryList;
