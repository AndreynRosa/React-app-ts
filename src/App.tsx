import React from 'react';
import { Provider } from 'react-redux';
import RepositoryList from './component/RepositoryList';
import store from './store';

const App: React.FC = () => (
  <Provider store={store}>
    <RepositoryList />
  </Provider>
);

export default App;
