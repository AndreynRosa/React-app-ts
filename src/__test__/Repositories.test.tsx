import React from 'react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import { shallow } from 'enzyme';
import moxios from 'moxios';
import createSagaMiddleware from 'redux-saga';
import { call } from 'redux-saga/effects';
import App from '../App';
import RepositoryList from '../component/RepositoryList';
import {
  repositoriesLoadRequest,
  repositoriesLoadSuccess,
} from '../store/ducks/repositories/actions';
import GetRepositoriesMock from './mock/GetRepositoriesMock';
import { loadRepositories } from '../store/ducks/repositories/sagas';
import { RepositoriesTypes } from '../store/ducks/repositories/types';

const initialState = {
  repositories: {
    data: [
      { name: 'Android-Java-Native-Exemple', id: 205602287 },
      { name: 'cultural-android-app', id: 216683402 },
      { name: 'Gerador-de-conteudos', id: 262713486 },
      { name: 'Java-Spring-and-React-', id: 245491399 },
      { name: 'Java-with-Junit-exemple', id: 237565428 },
    ],
    error: false,
    loading: false,
  },
};

// Configure store and pass sagamiddleware
const sagaMiddleware = createSagaMiddleware();
const mockStore = configureStore([sagaMiddleware]);
const store = mockStore(initialState);

// Component Test.
describe(' TESTING APP READIRING', () => {
  it('render correctly', () => {
    const warppert = shallow(<App />, { context: { store } });
    expect(warppert).toMatchSnapshot();
  });
});

describe(' TESTING REPOSITORY LIST COMPONENT', () => {
  it('render correctly', () => {
    const wrapper = shallow(
      <Provider store={store}>
        <RepositoryList />
      </Provider>,
    ).dive();
    expect(wrapper).toMatchSnapshot();
  });
});

// Actions Test
describe('ACTIONS FROM REPOSITORIES - TEST - GetRepositories ', () => {
  beforeEach(() => {
    moxios.install();
  });
  afterEach(() => {
    moxios.uninstall();
  });

  it('LOAD_REQUEST, LOAD_SUCCCES ', (done) => {
    moxios.wait(() => {
      const request = moxios.requests.mostRecent();
      request.respondWith({
        status: 200,
        response: { results: GetRepositoriesMock },
      });
    });

    const expectedActions = [repositoriesLoadRequest('AndreynRosa')];
    store.subscribe(() => {
      const actions = store.getActions();
      if (actions.length >= expectedActions.length) {
        expect(actions).toEqual(expectedActions);
        done();
      }
    });

    store.dispatch(repositoriesLoadRequest('AndreynRosa'));
  });
});

// Sagas test

describe('SAGAS - TEST - RepositoriesLoadSuccess ', () => {
  it('sucess actions with repositories', () => {
    const generator = loadRepositories({
      type: RepositoriesTypes.LOAD_REQUEST,
      payload: {
        name: '',
      },
    });
    const response = { data: { results: GetRepositoriesMock } };
    expect(generator.next(response).value).toBeNull;
  });
});
