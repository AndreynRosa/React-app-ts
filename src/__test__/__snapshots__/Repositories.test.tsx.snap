// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` TESTING APP READIRING render correctly 1`] = `ShallowWrapper {}`;

exports[` TESTING REPOSITORY LIST COMPONENT render correctly 1`] = `ShallowWrapper {}`;
